examples/parser/Parser/CSV.roc

```diff
@@ -0,0 +1,196 @@
+                |> Continue
+
+## Attempts to parse an `a` from a `CSVRecord` datastructure (a list of bytestring-fields)
+##
+## This parser succeeds when all fields of the CSVRecord are consumed by the parser.
+parseCSVRecord : Parser CSVRecord a, CSVRecord -> Result a [ParsingFailure Str, ParsingIncomplete CSVRecord]
+parseCSVRecord = \csvParser, recordFieldsList ->
+    parse csvParser recordFieldsList (\leftover -> leftover == [])
+
+## Wrapper function to combine a set of fields into your desired `a`
+##
+## ## Usage example
+##
+## >>> record (\firstName -> \lastName -> \age -> User {firstName, lastName, age})
+## >>> |> field string
+## >>> |> field string
+## >>> |> field nat
+##
+record : a -> Parser CSVRecord a
+record = Parser.Core.const
+
+## Turns a parser for a `List U8` into a parser that parses part of a `CSVRecord`.
+field : Parser RawStr a -> Parser CSVRecord a
+field = \fieldParser ->
+    buildPrimitiveParser \fieldsList ->
+        when List.get fieldsList 0 is
+            Err OutOfBounds ->
+                Err (ParsingFailure "expected another CSV field but there are no more fields in this record")
+
+            Ok rawStr ->
+                when Parser.Str.parseRawStr fieldParser rawStr is
+                    Ok val ->
+                        Ok { val: val, input: List.dropFirst fieldsList }
+
+                    Err (ParsingFailure reason) ->
+                        fieldStr = rawStr |> strFromRaw
+
+                        Err (ParsingFailure "Field `\(fieldStr)` could not be parsed. \(reason)")
+
+                    Err (ParsingIncomplete reason) ->
+                        reasonStr = strFromRaw reason
+                        fieldsStr = fieldsList |> List.map strFromRaw |> Str.joinWith ", "
+
+                        Err (ParsingFailure "The field parser was unable to read the whole field: `\(reasonStr)` while parsing the first field of leftover \(fieldsStr))")
+
+## Parser for a field containing a UTF8-encoded string
+string : Parser CSVField Str
+string = Parser.Str.anyString
+
+## Parse a natural number from a CSV field
+nat : Parser CSVField Nat
+nat =
+    string
+    |> map
+        (\val ->
+            when Str.toNat val is
+                Ok num ->
+                    Ok num
+
+                Err _ ->
+                    Err "\(val) is not a Nat."
+        )
+    |> flatten
+
+## Parse a 64-bit float from a CSV field
+f64 : Parser CSVField F64
+f64 =
+    string
+    |> map
+        (\val ->
+            when Str.toF64 val is
+                Ok num ->
+                    Ok num
+
+                Err _ ->
+                    Err "\(val) is not a F64."
+        )
+    |> flatten
+
+## Attempts to parse a Str into the internal `CSV` datastructure (A list of lists of bytestring-fields).
+parseStrToCSV : Str -> Result CSV [ParsingFailure Str, ParsingIncomplete RawStr]
+parseStrToCSV = \input ->
+    parse file (Str.toUtf8 input) (\leftover -> leftover == [])
```

Comment:
could also do

```suggestion
    parse file (Str.toUtf8 input) List.isEmpty
```

here and in `parseStrToCSVRecord`, which actually decreases binary size since only one function has to be created for the emptiness-checking, rather than two!